{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting bioframe\n",
      "  Downloading bioframe-0.2.0-py2.py3-none-any.whl (42 kB)\n",
      "\u001b[K     |████████████████████████████████| 42 kB 14 kB/s eta 0:00:012\n",
      "\u001b[?25hRequirement already satisfied: click>=7 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from bioframe) (8.0.3)\n",
      "Requirement already satisfied: requests in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from bioframe) (2.27.1)\n",
      "Requirement already satisfied: numpy>=1.10 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from bioframe) (1.19.5)\n",
      "Requirement already satisfied: pandas>=1.0.3 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from bioframe) (1.1.5)\n",
      "Requirement already satisfied: importlib-metadata in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from click>=7->bioframe) (4.8.3)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from pandas>=1.0.3->bioframe) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.2 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from pandas>=1.0.3->bioframe) (2022.4)\n",
      "Requirement already satisfied: six>=1.5 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from python-dateutil>=2.7.3->pandas>=1.0.3->bioframe) (1.16.0)\n",
      "Requirement already satisfied: typing-extensions>=3.6.4 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from importlib-metadata->click>=7->bioframe) (4.1.1)\n",
      "Requirement already satisfied: zipp>=0.5 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from importlib-metadata->click>=7->bioframe) (3.6.0)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from requests->bioframe) (1.26.8)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from requests->bioframe) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from requests->bioframe) (3.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from requests->bioframe) (2021.5.30)\n",
      "Installing collected packages: bioframe\n",
      "Successfully installed bioframe-0.2.0\n"
     ]
    }
   ],
   "source": [
    "!pip install bioframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting gtfparse\n",
      "  Downloading gtfparse-1.2.1.tar.gz (12 kB)\n",
      "Requirement already satisfied: numpy>=1.7 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from gtfparse) (1.19.5)\n",
      "Requirement already satisfied: pandas>=0.15 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from gtfparse) (1.1.5)\n",
      "Requirement already satisfied: pytz>=2017.2 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from pandas>=0.15->gtfparse) (2022.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from pandas>=0.15->gtfparse) (2.8.2)\n",
      "Requirement already satisfied: six>=1.5 in ./anaconda3/envs/protoqa/lib/python3.6/site-packages (from python-dateutil>=2.7.3->pandas>=0.15->gtfparse) (1.16.0)\n",
      "Building wheels for collected packages: gtfparse\n",
      "  Building wheel for gtfparse (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for gtfparse: filename=gtfparse-1.2.1-py3-none-any.whl size=11637 sha256=5c63e0f8bcb4557d381264f5665b169ab94ab1a71c2448963daa9b58f5666734\n",
      "  Stored in directory: /home1/kadvani/.cache/pip/wheels/3c/99/0d/8f7f3dda491b2212eb6ab4bf815a7dd8bd5926e10de353820a\n",
      "Successfully built gtfparse\n",
      "Installing collected packages: gtfparse\n",
      "Successfully installed gtfparse-1.2.1\n"
     ]
    }
   ],
   "source": [
    "!pip install gtfparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioframe\n",
    "import gtfparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home1/kadvani\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gene_name', 'exon_number', 'exon_id']\n"
     ]
    }
   ],
   "source": [
    "df = gtfparse.read_gtf(\n",
    "    '3dfolding/mm9.refGene.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15</td>\n",
       "      <td>refGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>76909420</td>\n",
       "      <td>77137483</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>NM_001110827</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>refGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>76909420</td>\n",
       "      <td>76914636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>NM_001110827</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>1</td>\n",
       "      <td>NM_001110827.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr15</td>\n",
       "      <td>refGene</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>76909420</td>\n",
       "      <td>76914591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>NM_001110827</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>1</td>\n",
       "      <td>NM_001110827.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr15</td>\n",
       "      <td>refGene</td>\n",
       "      <td>CDS</td>\n",
       "      <td>76914595</td>\n",
       "      <td>76914636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>NM_001110827</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>1</td>\n",
       "      <td>NM_001110827.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr15</td>\n",
       "      <td>refGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>76916008</td>\n",
       "      <td>76916080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>NM_001110827</td>\n",
       "      <td>Rbfox2</td>\n",
       "      <td>2</td>\n",
       "      <td>NM_001110827.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012735</th>\n",
       "      <td>chr19</td>\n",
       "      <td>refGene</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>24268199</td>\n",
       "      <td>24268319</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>NM_001376368</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>23</td>\n",
       "      <td>NM_001376368.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012736</th>\n",
       "      <td>chr19</td>\n",
       "      <td>refGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>24299341</td>\n",
       "      <td>24299516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>NM_001376368</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>24</td>\n",
       "      <td>NM_001376368.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012737</th>\n",
       "      <td>chr19</td>\n",
       "      <td>refGene</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>24299341</td>\n",
       "      <td>24299516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>NM_001376368</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>24</td>\n",
       "      <td>NM_001376368.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012738</th>\n",
       "      <td>chr19</td>\n",
       "      <td>refGene</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>24213311</td>\n",
       "      <td>24213313</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>NM_001376368</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>22</td>\n",
       "      <td>NM_001376368.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012739</th>\n",
       "      <td>chr19</td>\n",
       "      <td>refGene</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>24169915</td>\n",
       "      <td>24169917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>NM_001376368</td>\n",
       "      <td>Tjp2</td>\n",
       "      <td>1</td>\n",
       "      <td>NM_001376368.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1012740 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source      feature     start       end  score strand frame  \\\n",
       "0         chr15  refGene   transcript  76909420  77137483    NaN      -     0   \n",
       "1         chr15  refGene         exon  76909420  76914636    NaN      -     0   \n",
       "2         chr15  refGene         3UTR  76909420  76914591    NaN      -     0   \n",
       "3         chr15  refGene          CDS  76914595  76914636    NaN      -     0   \n",
       "4         chr15  refGene         exon  76916008  76916080    NaN      -     0   \n",
       "...         ...      ...          ...       ...       ...    ...    ...   ...   \n",
       "1012735   chr19  refGene         5UTR  24268199  24268319    NaN      -     0   \n",
       "1012736   chr19  refGene         exon  24299341  24299516    NaN      -     0   \n",
       "1012737   chr19  refGene         5UTR  24299341  24299516    NaN      -     0   \n",
       "1012738   chr19  refGene  start_codon  24213311  24213313    NaN      -     0   \n",
       "1012739   chr19  refGene   stop_codon  24169915  24169917    NaN      -     0   \n",
       "\n",
       "        gene_id transcript_id gene_name exon_number          exon_id  \n",
       "0        Rbfox2  NM_001110827    Rbfox2                               \n",
       "1        Rbfox2  NM_001110827    Rbfox2           1   NM_001110827.1  \n",
       "2        Rbfox2  NM_001110827    Rbfox2           1   NM_001110827.1  \n",
       "3        Rbfox2  NM_001110827    Rbfox2           1   NM_001110827.1  \n",
       "4        Rbfox2  NM_001110827    Rbfox2           2   NM_001110827.2  \n",
       "...         ...           ...       ...         ...              ...  \n",
       "1012735    Tjp2  NM_001376368      Tjp2          23  NM_001376368.23  \n",
       "1012736    Tjp2  NM_001376368      Tjp2          24  NM_001376368.24  \n",
       "1012737    Tjp2  NM_001376368      Tjp2          24  NM_001376368.24  \n",
       "1012738    Tjp2  NM_001376368      Tjp2          22  NM_001376368.22  \n",
       "1012739    Tjp2  NM_001376368      Tjp2           1   NM_001376368.1  \n",
       "\n",
       "[1012740 rows x 13 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
