{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n",
      "\n",
      "CondaError: KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda install -c bioconda pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gffutils\n",
    "import pybedtools\n",
    "from pybedtools.featurefuncs import TSS\n",
    "from gffutils.helpers import asinterval\n",
    "import bioframe as bf\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gene_type', 'gene_name', 'transcript_type', 'transcript_name', 'level', 'transcript_support_level', 'mgi_id', 'tag', 'havana_gene', 'havana_transcript', 'protein_id', 'ccdsid', 'ont']\n"
     ]
    }
   ],
   "source": [
    "tss_df = read_gtf(\"/project/fudenber_735/collaborations/karissa_2022/old/RNAseq/STAR_Gencode_alignment/tss_annotions_gencode.vM23.primary_assembly.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3073252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>...</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>ENSMUST00000082908.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>MGI:5455983</td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3216345</td>\n",
       "      <td>3216345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUST00000162897.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Xkr4-203</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>OTTMUST00000086625.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3215633</td>\n",
       "      <td>3215633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUST00000159265.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Xkr4-202</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>OTTMUST00000086624.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3671499</td>\n",
       "      <td>3671499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "      <td>...</td>\n",
       "      <td>Xkr4-201</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>OTTMUST00000065166.1</td>\n",
       "      <td>ENSMUSP00000070648.4</td>\n",
       "      <td>CCDS14803.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142441</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>913084</td>\n",
       "      <td>913084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095475.1</td>\n",
       "      <td>ENSMUST00000180208.1</td>\n",
       "      <td>...</td>\n",
       "      <td>AC133095.2-201</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000137110.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142442</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>921941</td>\n",
       "      <td>921941</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000094855.1</td>\n",
       "      <td>ENSMUST00000178327.1</td>\n",
       "      <td>...</td>\n",
       "      <td>AC133095.1-201</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000137276.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142443</th>\n",
       "      <td>JH584303.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>81606</td>\n",
       "      <td>81606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095019.1</td>\n",
       "      <td>ENSMUST00000178569.1</td>\n",
       "      <td>...</td>\n",
       "      <td>AC234645.1-201</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td></td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000136543.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142444</th>\n",
       "      <td>JH584304.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>59668</td>\n",
       "      <td>59668</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095041.7</td>\n",
       "      <td>ENSMUST00000179505.7</td>\n",
       "      <td>...</td>\n",
       "      <td>AC149090.1-201</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000136028.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142445</th>\n",
       "      <td>JH584304.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>59691</td>\n",
       "      <td>59691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095041.7</td>\n",
       "      <td>ENSMUST00000178343.1</td>\n",
       "      <td>...</td>\n",
       "      <td>AC149090.1-202</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000136649.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142446 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           seqname   source     feature    start      end  score strand frame  \\\n",
       "0             chr1   HAVANA  transcript  3073252  3073252    NaN      +     0   \n",
       "1             chr1  ENSEMBL  transcript  3102015  3102015    NaN      +     0   \n",
       "2             chr1   HAVANA  transcript  3216345  3216345    NaN      -     0   \n",
       "3             chr1   HAVANA  transcript  3215633  3215633    NaN      -     0   \n",
       "4             chr1   HAVANA  transcript  3671499  3671499    NaN      -     0   \n",
       "...            ...      ...         ...      ...      ...    ...    ...   ...   \n",
       "142441  JH584299.1  ENSEMBL  transcript   913084   913084    NaN      -     0   \n",
       "142442  JH584299.1  ENSEMBL  transcript   921941   921941    NaN      +     0   \n",
       "142443  JH584303.1  ENSEMBL  transcript    81606    81606    NaN      +     0   \n",
       "142444  JH584304.1  ENSEMBL  transcript    59668    59668    NaN      -     0   \n",
       "142445  JH584304.1  ENSEMBL  transcript    59691    59691    NaN      -     0   \n",
       "\n",
       "                     gene_id         transcript_id  ...    transcript_name  \\\n",
       "0       ENSMUSG00000102693.1  ENSMUST00000193812.1  ...  4933401J01Rik-201   \n",
       "1       ENSMUSG00000064842.1  ENSMUST00000082908.1  ...        Gm26206-201   \n",
       "2       ENSMUSG00000051951.5  ENSMUST00000162897.1  ...           Xkr4-203   \n",
       "3       ENSMUSG00000051951.5  ENSMUST00000159265.1  ...           Xkr4-202   \n",
       "4       ENSMUSG00000051951.5  ENSMUST00000070533.4  ...           Xkr4-201   \n",
       "...                      ...                   ...  ...                ...   \n",
       "142441  ENSMUSG00000095475.1  ENSMUST00000180208.1  ...     AC133095.2-201   \n",
       "142442  ENSMUSG00000094855.1  ENSMUST00000178327.1  ...     AC133095.1-201   \n",
       "142443  ENSMUSG00000095019.1  ENSMUST00000178569.1  ...     AC234645.1-201   \n",
       "142444  ENSMUSG00000095041.7  ENSMUST00000179505.7  ...     AC149090.1-201   \n",
       "142445  ENSMUSG00000095041.7  ENSMUST00000178343.1  ...     AC149090.1-202   \n",
       "\n",
       "       level transcript_support_level       mgi_id  \\\n",
       "0          2                       NA  MGI:1918292   \n",
       "1          3                       NA  MGI:5455983   \n",
       "2          2                        1  MGI:3528744   \n",
       "3          2                        1  MGI:3528744   \n",
       "4          2                        1  MGI:3528744   \n",
       "...      ...                      ...          ...   \n",
       "142441     3                        1                \n",
       "142442     3                        5                \n",
       "142443     3                       NA                \n",
       "142444     3                        1                \n",
       "142445     3                        1                \n",
       "\n",
       "                                  tag           havana_gene  \\\n",
       "0                               basic  OTTMUSG00000049935.1   \n",
       "1                               basic                         \n",
       "2                                      OTTMUSG00000026353.2   \n",
       "3                                      OTTMUSG00000026353.2   \n",
       "4       basic,appris_principal_1,CCDS  OTTMUSG00000026353.2   \n",
       "...                               ...                   ...   \n",
       "142441                          basic                         \n",
       "142442                          basic                         \n",
       "142443                          basic                         \n",
       "142444                          basic                         \n",
       "142445                          basic                         \n",
       "\n",
       "           havana_transcript            protein_id       ccdsid ont  \n",
       "0       OTTMUST00000127109.1                                         \n",
       "1                                                                    \n",
       "2       OTTMUST00000086625.1                                         \n",
       "3       OTTMUST00000086624.1                                         \n",
       "4       OTTMUST00000065166.1  ENSMUSP00000070648.4  CCDS14803.1      \n",
       "...                      ...                   ...          ...  ..  \n",
       "142441                        ENSMUSP00000137110.1                   \n",
       "142442                        ENSMUSP00000137276.1                   \n",
       "142443                        ENSMUSP00000136543.1                   \n",
       "142444                        ENSMUSP00000136028.1                   \n",
       "142445                        ENSMUSP00000136649.1                   \n",
       "\n",
       "[142446 rows x 23 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cleaning out less-well defined chromosome numbers\n",
    "tss_df = tss_df.loc[False==( tss_df['seqname'].str.contains('NT_'))]\n",
    "tss_df = tss_df.loc[False==( tss_df['seqname'].str.contains('MT'))]\n",
    "\n",
    "# paste 'chr' to all chromosome names\n",
    "tss_df[\"seqname\"] = tss_df[\"seqname\"]\n",
    "\n",
    "# rename column to chrom to match bedframe/bioframe format\n",
    "tss_df = tss_df.rename(columns= {\"seqname\" : \"chrom\"})\n",
    "\n",
    "# Removing pseudo chromosomes\n",
    "before = len(tss_df)\n",
    "tss_df = tss_df.loc[False==( tss_df['chrom'].str.contains('chrGL'))]\n",
    "tss_df = tss_df.loc[False==( tss_df['chrom'].str.contains('chrJH'))]\n",
    "tss_df = tss_df.loc[False==( tss_df['chrom'].str.contains('chrX'))]\n",
    "tss_df = tss_df.loc[False==( tss_df['chrom'].str.contains('chrY'))]\n",
    "tss_df = tss_df.loc[False==( tss_df['chrom'].str.contains('chrM'))]\n",
    "tss_df =tss_df.loc[True==tss_df['chrom'].str.contains('chr')]\n",
    "after = len(tss_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_df['chrom'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, source, feature, start, end, score, strand, frame, gene_id, transcript_id, gene_type, gene_name, transcript_type, transcript_name, level, transcript_support_level, mgi_id, tag, havana_gene, havana_transcript, protein_id, ccdsid, ont]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 23 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_df[False == (tss_df['chrom'].str.contains(\"chr\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop duplicates (same across all columns we are interested in for defining intervals)\n",
    "ts_intervals = tss_df[['gene_id','chrom', 'start', 'end']].drop_duplicates(['gene_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3073252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3216345</td>\n",
       "      <td>3216345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000102851.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3252756</td>\n",
       "      <td>3252756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000103377.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3368550</td>\n",
       "      <td>3368550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134932</th>\n",
       "      <td>ENSMUSG00000117740.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61232927</td>\n",
       "      <td>61232927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134933</th>\n",
       "      <td>ENSMUSG00000117782.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61255966</td>\n",
       "      <td>61255966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134934</th>\n",
       "      <td>ENSMUSG00000117951.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61257790</td>\n",
       "      <td>61257790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134935</th>\n",
       "      <td>ENSMUSG00000095993.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61297070</td>\n",
       "      <td>61297070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134936</th>\n",
       "      <td>ENSMUSG00000118197.1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61305281</td>\n",
       "      <td>61305281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51110 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id  chrom     start       end\n",
       "0       ENSMUSG00000102693.1   chr1   3073252   3073252\n",
       "1       ENSMUSG00000064842.1   chr1   3102015   3102015\n",
       "2       ENSMUSG00000051951.5   chr1   3216345   3216345\n",
       "5       ENSMUSG00000102851.1   chr1   3252756   3252756\n",
       "6       ENSMUSG00000103377.1   chr1   3368550   3368550\n",
       "...                      ...    ...       ...       ...\n",
       "134932  ENSMUSG00000117740.1  chr19  61232927  61232927\n",
       "134933  ENSMUSG00000117782.1  chr19  61255966  61255966\n",
       "134934  ENSMUSG00000117951.1  chr19  61257790  61257790\n",
       "134935  ENSMUSG00000095993.1  chr19  61297070  61297070\n",
       "134936  ENSMUSG00000118197.1  chr19  61305281  61305281\n",
       "\n",
       "[51110 rows x 4 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
