{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "881fe14f-e758-4108-bc95-c0cb75361b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import bioframe \n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import os\n",
    "import json\n",
    "from io import StringIO\n",
    "\n",
    "import akita_utils \n",
    "import pysam\n",
    "import h5py\n",
    "genome_open = pysam.Fastafile('/project/fudenber_735/genomes/mm10/mm10.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "40a0bf0d-860a-4b3c-9ef9-13dd4f319b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.6.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-06-16 17:40:06.076908: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /spack/apps/linux-centos7-x86_64/gcc-8.3.0/python-3.9.2-uvcroioc4witkp6qf7mbebof5ix4wlb6/lib:/spack/apps/linux-centos7-x86_64/gcc-8.3.0/pmix-3.1.3-3sm6emyqaxapunh7rwbjvtaqoqe2e5z3/lib:/spack/apps/linux-centos7-x86_64/gcc-8.3.0/openmpi-4.0.2-ipm3dnvlbtxawpi4ifz7jma6jgr7mexq/lib:/spack/apps/linux-centos7-x86_64/gcc-8.3.0/openblas-0.3.8-2no6mfziiclwxb7lstxoos335gnhjpes/lib:/spack/apps/gcc/8.3.0/lib64\n",
      "2022-06-16 17:40:06.076946: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)\n",
      "2022-06-16 17:40:06.076970: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (d05-43.hpc.usc.edu): /proc/driver/nvidia/version does not exist\n",
      "2022-06-16 17:40:06.077290: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "built\n",
      "restored\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(512, 130305, (130305,))"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = '-1'\n",
    "import tensorflow as tf\n",
    "print(tf.__version__)\n",
    "\n",
    "import pysam\n",
    "from basenji import dataset, seqnn, dna_io,stream\n",
    "head_i = 1 #mouse\n",
    "model_num = 1 #which fold to use\n",
    "\n",
    "base_dir = '/project/fudenber_735/tensorflow_models/akita/v2/models/'\n",
    "model_dir = base_dir+\"/f\"+str(model_num)+\"c0/train/\"\n",
    "model_file  = model_dir+'/model'+str(head_i)+'_best.h5'\n",
    "\n",
    "params_file = model_dir+'/params.json'\n",
    "with open(params_file) as params_open:\n",
    "    params = json.load(params_open)\n",
    "    params_model = params['model']\n",
    "    params_train = params['train']\n",
    "seq_length = params_model['seq_length']\n",
    "params_model['verbose'] = False\n",
    "seqnn_model = seqnn.SeqNN(params_model)\n",
    "print('built')\n",
    "seqnn_model.restore(model_file, head_i=head_i)\n",
    "print('restored')\n",
    "\n",
    "hic_diags = params_model['diagonal_offset']\n",
    "try:\n",
    "    target_crop = params_model['trunk'][-2]['cropping']\n",
    "except:\n",
    "    target_crop = params_model['target_crop']\n",
    "target_length_cropped = int((seq_length//2048 - target_crop*2 - hic_diags)* ((seq_length//2048 - target_crop*2 - hic_diags) +1)/2) \n",
    "target_map_size = seq_length//2048  - target_crop*2 \n",
    "triu_tup = np.triu_indices(target_map_size,2)\n",
    "target_map_size, target_length_cropped, triu_tup[0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3d71294a-57f4-4dbb-a042-598f19e62497",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29 duplicates removed for  /project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model7/scd.h5\n",
      "29 duplicates removed for  /project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model1/scd.h5\n",
      "annotating each site with boundary-wide scores\n",
      "filtering sites by overlap with rmsk\n",
      "df prepared\n"
     ]
    }
   ],
   "source": [
    "site_df = akita_utils.prepare_insertion_tsv(\n",
    "    h5_dirs = '/project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model*/scd.h5',\n",
    "    score_key = 'SCD',\n",
    "    flank_pad = 60, #how much flanking sequence around the sites to include\n",
    "    weak_thresh_pct = 1, # don't use sites weaker than this, might be artifacts\n",
    "    weak_num = 5 ,\n",
    "    strong_thresh_pct = 99, # don't use sites weaker than this, might be artifacts\n",
    "    strong_num = 5 ,\n",
    "    save_tsv=None, # optional filename to save a tsv\n",
    ")\n",
    "num_seqs = site_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d1791ab6-36e8-4914-a243-dc93e163f65a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>genomic_SCD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr12</td>\n",
       "      <td>35192299</td>\n",
       "      <td>35192438</td>\n",
       "      <td>-</td>\n",
       "      <td>53.656250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>102781052</td>\n",
       "      <td>102781191</td>\n",
       "      <td>-</td>\n",
       "      <td>53.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr5</td>\n",
       "      <td>49961931</td>\n",
       "      <td>49962070</td>\n",
       "      <td>-</td>\n",
       "      <td>53.312500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>26776296</td>\n",
       "      <td>26776435</td>\n",
       "      <td>-</td>\n",
       "      <td>53.125000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr9</td>\n",
       "      <td>51152529</td>\n",
       "      <td>51152668</td>\n",
       "      <td>-</td>\n",
       "      <td>53.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0</td>\n",
       "      <td>chr7</td>\n",
       "      <td>139598251</td>\n",
       "      <td>139598390</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>45342859</td>\n",
       "      <td>45342998</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27771606</td>\n",
       "      <td>27771745</td>\n",
       "      <td>-</td>\n",
       "      <td>0.058746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>35713828</td>\n",
       "      <td>35713967</td>\n",
       "      <td>+</td>\n",
       "      <td>0.059021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4</td>\n",
       "      <td>chr9</td>\n",
       "      <td>51333550</td>\n",
       "      <td>51333689</td>\n",
       "      <td>-</td>\n",
       "      <td>0.059021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  chrom      start        end strand  genomic_SCD\n",
       "0      0  chr12   35192299   35192438      -    53.656250\n",
       "1      1   chr8  102781052  102781191      -    53.500000\n",
       "2      2   chr5   49961931   49962070      -    53.312500\n",
       "3      3   chr9   26776296   26776435      -    53.125000\n",
       "4      4   chr9   51152529   51152668      -    53.062500\n",
       "5      0   chr7  139598251  139598390      +     0.058136\n",
       "6      1   chr8   45342859   45342998      +     0.058319\n",
       "7      2   chr2   27771606   27771745      -     0.058746\n",
       "8      3   chr1   35713828   35713967      +     0.059021\n",
       "9      4   chr9   51333550   51333689      -     0.059021"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_df#['end']- site_df['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d03fb1cd-4243-4e91-ab8c-ca9710580a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "ind 0 , iter  0 , for chr12 34537008 35847728\n",
      "0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-06-16 17:42:16.798224: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "success: best seq, thresh 4676.0  pixelwise 0.03406\n",
      "max iters exceeded\n",
      "1\n",
      "ind 1 , iter  0 , for chr8 102125761 103436481\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "max_iters = 1\n",
    "batch_size = 6\n",
    "shuffle_k = 8\n",
    "\n",
    "## alternate approach, figure out a way to nicely use the whole JASPAR motif\n",
    "# motif_file = '/project/fudenber_735/motifs/pfms/JASPAR2022_CORE_redundant_pfms_jaspar/MA0139.1.jaspar'\n",
    "# motif = read_jaspar_to_numpy(motif_file)\n",
    "### using just the core motif and this shuffling which also makes the reverse complement a low score works well\n",
    "mot = '>CCAsyAGrkGGCr\\n0.0000\\t1.0000\\t0.0000\\t0.0000\\n0.0000\\t1.0000\\t0.0000\\t0.0000\\n1.0000\\t0.0000\\t0.0000\\t0.0000\\n0.0000\\t0.5000\\t0.5000\\t0.0000\\n0.0000\\t0.5000\\t0.0000\\t0.5000\\n1.0000\\t0.0000\\t0.0000\\t0.0000\\n0.0000\\t0.0000\\t1.0000\\t0.0000\\n0.5000\\t0.0000\\t0.5000\\t0.0000\\n0.0000\\t0.0000\\t0.5000\\t0.5000\\n0.0000\\t0.0000\\t1.0000\\t0.0000\\n0.0000\\t0.0000\\t1.0000\\t0.0000\\n0.0000\\t1.0000\\t0.0000\\t0.0000\\n0.5000\\t0.0000\\t0.5000\\t0.0000'\n",
    "motif = pd.read_csv(StringIO(mot),sep='\\t',header=0, names=['A','C','G','T']).values\n",
    "motif_window = int(np.ceil(len(motif)/2))\n",
    "mot_shuf = np.array([12,0,1,11,10,3,2,8,9,4,5,7,6]) \n",
    "\n",
    "ctcf_thresh = 8\n",
    "scores_thresh = 5500\n",
    "scores_pixelwise_thresh = .04\n",
    "\n",
    "\n",
    "\n",
    "flat_seqs = []\n",
    "for ind in range(num_seqs):\n",
    "    print(ind)\n",
    "    chrom, start, end, strand = site_df.iloc[ind][['chrom','start','end','strand']]\n",
    "    mid = int(.5 * (start +end))\n",
    "    start, end =  mid - seq_length//2 , mid + seq_length//2\n",
    "    seq = genome_open.fetch(chrom, start, end).upper()\n",
    "    #if strand == '-': seq = dna_rc(seq)\n",
    "    seq_1hot = dna_io.dna_1hot(seq)\n",
    "\n",
    "    num_iters = 0\n",
    "    while num_iters < max_iters:\n",
    "        print('ind',ind, ', iter ',num_iters,', for', chrom, start, end)\n",
    "        print(len(flat_seqs))\n",
    "        \n",
    "        seq_1hot_batch = []\n",
    "        for i in range(batch_size):\n",
    "            seq_1hot_mut = akita_utils.permute_seq_k(seq_1hot,k= shuffle_k)\n",
    "            s = akita_utils.scan_motif(seq_1hot_mut, motif  )\n",
    "            for i in np.where(s > ctcf_thresh)[0]:\n",
    "                #seq_1hot_mut[i-motif_window:i+motif_window] = permute_seq_k(seq_1hot_mut[i-motif_window:i+motif_window], k=2)\n",
    "                seq_1hot_mut[i-motif_window+1:i+motif_window] = seq_1hot_mut[i-motif_window+1:i+motif_window][mot_shuf]\n",
    "            seq_1hot_batch.append(seq_1hot_mut)\n",
    "        seq_1hot_batch = np.array(seq_1hot_batch)\n",
    "\n",
    "        pred = seqnn_model.predict(seq_1hot_batch, batch_size=batch_size)\n",
    "        scores = np.sum( pred**2, axis=-1).sum(axis=-1)\n",
    "        scores_pixelwise = np.max(pred**2, axis=-1).max(axis=-1)\n",
    "\n",
    "        if np.any( (np.min(scores) < scores_thresh) * (np.min(scores_pixelwise) < scores_pixelwise_thresh)):\n",
    "            best_ind = np.argmin(scores_pixelwise)\n",
    "            best_seq = seq_1hot_batch[best_ind]\n",
    "            best_pred = pred[best_ind]\n",
    "            best_score, best_score_pixelwise = scores[best_ind] , scores_pixelwise[best_ind]\n",
    "            num_iters = max_iters\n",
    "            print('success: best seq, thresh', np.min(scores),' pixelwise',np.min(scores_pixelwise))\n",
    "\n",
    "        else: \n",
    "            best_ind = np.argmin(scores_pixelwise)\n",
    "            best_seq = seq_1hot_batch[best_ind]\n",
    "            best_pred = pred[best_ind]\n",
    "            best_score, best_score_pixelwise = scores[best_ind] , scores_pixelwise[best_ind]\n",
    "            print('trying: best seq, thresh', np.min(scores),' pixelwise',np.min(scores_pixelwise))\n",
    "\n",
    "        num_iters +=1\n",
    "        if num_iters >= max_iters:\n",
    "            print('max iters exceeded')\n",
    "            flat_seqs.append( [best_seq, best_pred, best_score, best_score_pixelwise] )\n",
    "            #raise ValueError('cannot generate flat sequence for', chrom, start, end)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d748b91d-7213-44a5-905e-5fd11c1494b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8,8))#,6))\n",
    "\n",
    "target_ind = 0\n",
    "vlim = 1.5\n",
    "\n",
    "for i in range(len(flat_seqs)):\n",
    "    flat_pred = flat_seqs[i][1]\n",
    "    \n",
    "    plt.subplot(3,4, i+1)\n",
    "    im = plt.matshow(\n",
    "            from_upper_triu(  flat_pred[:,target_ind], target_map_size,hic_diags),\n",
    "             vmin=-1*vlim,vmax=vlim, fignum=False,cmap='RdBu_r')\n",
    "    plt.colorbar(im, fraction=0.046, pad=0.04)\n",
    "    plt.title('tot '+\n",
    "            str(np.round(flat_seqs[i][2],0).astype(int))+'\\n pixel '+\n",
    "            str(np.round(flat_seqs[i][3],4)) ) \n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d9cfec-cc89-43c6-985c-ba4151c638d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_panels = batch_size\n",
    "#plt.figure(figsize=(5*2,2*num_panels))\n",
    "plt.figure(figsize=(8,8))#,6))\n",
    "\n",
    "target_ind = 0\n",
    "vlim = 1.5\n",
    "\n",
    "for i in range(len(flat_seqs)):\n",
    "    flat_pred = flat_seqs[i][1]\n",
    "    \n",
    "    plt.subplot(3,4, i+1)\n",
    "    im = plt.matshow(\n",
    "            from_upper_triu(  flat_pred[:,target_ind], target_map_size,hic_diags),\n",
    "             vmin=-1*vlim,vmax=vlim, fignum=False,cmap='RdBu_r')\n",
    "    plt.colorbar(im, fraction=0.046, pad=0.04)\n",
    "    plt.title('tot '+\n",
    "            str(np.round(flat_seqs[i][2],0).astype(int))+'\\n pixel '+\n",
    "            str(np.round(flat_seqs[i][3],4)) ) \n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9065f6c6-74ea-4698-811b-51d294c68966",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('backround_seqs.fa','w') as f:\n",
    "    for i in range(len(flat_seqs)):\n",
    "        f.write('>seq'+str(i)+'_score'+str(int(flat_seqs[i][2]))+'_pixelwise'+str(int(flat_seqs[i][3]*1000))+'\\n')\n",
    "        f.write(dna_io.hot1_dna(flat_seqs[i][0])+'\\n')\n",
    "       "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basenji-geoff",
   "language": "python",
   "name": "basenji-geoff"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
